### ISSUE TITLE
Fails generation for List<VoidCallback>.

### ISSUE THREAD

**Issue by polina-c** [dart-lang/mockito#579]
For getter like this:
```
List<VoidCallback> get listeners => _listeners;
```

Generates getter:
```
List<dynamic> ...
```

that gives compile error:
`
'MockDebuggerController.listeners' ('List<dynamic> Function()') isn't a valid override of 'AutoDisposeControllerMixin.listeners' ('List<void Function()> Function()').
`

Replacement of `VoidCallback` with `void Function()` fixes the issue: https://github.com/flutter/devtools/pull/4641

**Comment by srawlins** [comment]
Interesting! Can you give a complete reproduction? With the declaration of `VoidCallback` please

**Comment by jpnurmi** [comment]
This started happening with e.g. `ChangeNotifier` from [flutter/foundation.dart](https://github.com/flutter/flutter/blob/d9a2229babbc7e28cabd0ad42470c527a6cd9d3e/packages/flutter/lib/src/foundation/change_notifier.dart#L112).

pubspec.yaml
```yaml
dev_dependencies:
  mockito: ^5.3.2
  build_runner: ^2.3.2
```

widget_test.dart
```dart
import 'package:flutter/foundation.dart';
import 'package:mockito/annotations.dart';

import 'widget_test.mocks.dart';

@GenerateMocks([ChangeNotifier])
void main() {}
```

widget_test.mocks.dart
```dart
// Mocks generated by Mockito 5.3.2 from annotations
// in mockito_app/test/widget_test.dart.
// Do not manually edit this file.

// ignore_for_file: no_leading_underscores_for_library_prefixes
import 'package:flutter/src/foundation/change_notifier.dart' as _i2;
import 'package:mockito/mockito.dart' as _i1;

// ignore_for_file: type=lint
// ignore_for_file: avoid_redundant_argument_values
// ignore_for_file: avoid_setters_without_getters
// ignore_for_file: comment_references
// ignore_for_file: implementation_imports
// ignore_for_file: invalid_use_of_visible_for_testing_member
// ignore_for_file: prefer_const_constructors
// ignore_for_file: unnecessary_parenthesis
// ignore_for_file: camel_case_types
// ignore_for_file: subtype_of_sealed_class

/// A class which mocks [ChangeNotifier].
///
/// See the documentation for Mockito's code generation for more information.
class MockChangeNotifier extends _i1.Mock implements _i2.ChangeNotifier {
  MockChangeNotifier() {
    _i1.throwOnMissingStub(this);
  }

  @override
  bool get hasListeners => (super.noSuchMethod(
        Invocation.getter(#hasListeners),
        returnValue: false,
      ) as bool);
  @override
  void addListener(dynamic listener) => super.noSuchMethod(
        Invocation.method(
          #addListener,
          [listener],
        ),
        returnValueForMissingStub: null,
      );
  @override
  void removeListener(dynamic listener) => super.noSuchMethod(
        Invocation.method(
          #removeListener,
          [listener],
        ),
        returnValueForMissingStub: null,
      );
  @override
  void dispose() => super.noSuchMethod(
        Invocation.method(
          #dispose,
          [],
        ),
        returnValueForMissingStub: null,
      );
  @override
  void notifyListeners() => super.noSuchMethod(
        Invocation.method(
          #notifyListeners,
          [],
        ),
        returnValueForMissingStub: null,
      );
}
```

**Comment by jpnurmi** [comment]
I went back to check the `flutter pub upgrade` output.
```
Resolving dependencies...
> _fe_analyzer_shared 50.0.0 (was 49.0.0)
> analyzer 5.2.0 (was 5.1.0)
  args 2.3.1
  async 2.9.0 (2.10.0 available)
  boolean_selector 2.1.0 (2.1.1 available)
  build 2.3.1
  build_config 1.1.1
  build_daemon 3.1.0
  build_resolvers 2.0.10
  build_runner 2.3.2
> build_runner_core 7.2.7 (was 7.2.6)
  built_collection 5.1.1
  built_value 8.4.1
  characters 1.2.1
  checked_yaml 2.0.1
  clock 1.1.1
  code_builder 4.3.0
  collection 1.16.0 (1.17.0 available)
  convert 3.1.1
  crypto 3.0.2
  dart_style 2.2.4
  fake_async 1.3.1
  file 6.1.4
  fixnum 1.0.1
  flutter 0.0.0 from sdk flutter
  flutter_lints 2.0.1
  flutter_svg 1.1.6
  flutter_test 0.0.0 from sdk flutter
  frontend_server_client 3.1.0
  glob 2.1.0
  graphs 2.2.0
  http_multi_server 3.2.1
  http_parser 4.0.2
  io 1.0.3
  js 0.6.5
  json_annotation 4.7.0
  lints 2.0.1
  logging 1.1.0
  matcher 0.12.12 (0.12.13 available)
  material_color_utilities 0.1.5 (0.2.0 available)
  meta 1.8.0
  mime 1.0.2
  mockito 5.3.2
  package_config 2.1.0
  path 1.8.2
  path_drawing 1.0.1
  path_parsing 1.0.1
  petitparser 5.1.0
  pool 1.5.1
  pub_semver 2.1.2
  pubspec_parse 1.2.1
  shelf 1.4.0
  shelf_web_socket 1.0.2
  sky_engine 0.0.99 from sdk flutter
  source_gen 1.2.6
  source_span 1.9.0 (1.9.1 available)
  stack_trace 1.10.0 (1.11.0 available)
  stream_channel 2.1.0 (2.1.1 available)
  stream_transform 2.0.1
  string_scanner 1.1.1
  term_glyph 1.2.1
  test_api 0.4.12 (0.4.15 available)
  timing 1.0.0
  typed_data 1.3.1
  vector_math 2.1.2 (2.1.4 available)
  watcher 1.0.2
  web_socket_channel 2.2.0
  xml 6.1.0 (6.2.0 available)
  yaml 3.1.1
```
I tried downgrading to analyzer 5.1.0:
```yaml
dependency_overrides:
  analyzer: 5.1.0
```
After that, the mock is generated fine:
```dart
// Mocks generated by Mockito 5.3.2 from annotations
// in mockito_app/test/widget_test.dart.
// Do not manually edit this file.

// ignore_for_file: no_leading_underscores_for_library_prefixes
import 'dart:ui' as _i3;

import 'package:flutter/src/foundation/change_notifier.dart' as _i2;
import 'package:mockito/mockito.dart' as _i1;

// ignore_for_file: type=lint
// ignore_for_file: avoid_redundant_argument_values
// ignore_for_file: avoid_setters_without_getters
// ignore_for_file: comment_references
// ignore_for_file: implementation_imports
// ignore_for_file: invalid_use_of_visible_for_testing_member
// ignore_for_file: prefer_const_constructors
// ignore_for_file: unnecessary_parenthesis
// ignore_for_file: camel_case_types
// ignore_for_file: subtype_of_sealed_class

/// A class which mocks [ChangeNotifier].
///
/// See the documentation for Mockito's code generation for more information.
class MockChangeNotifier extends _i1.Mock implements _i2.ChangeNotifier {
  MockChangeNotifier() {
    _i1.throwOnMissingStub(this);
  }

  @override
  bool get hasListeners => (super.noSuchMethod(
        Invocation.getter(#hasListeners),
        returnValue: false,
      ) as bool);
  @override
  void addListener(_i3.VoidCallback? listener) => super.noSuchMethod(
        Invocation.method(
          #addListener,
          [listener],
        ),
        returnValueForMissingStub: null,
      );
  @override
  void removeListener(_i3.VoidCallback? listener) => super.noSuchMethod(
        Invocation.method(
          #removeListener,
          [listener],
        ),
        returnValueForMissingStub: null,
      );
  @override
  void dispose() => super.noSuchMethod(
        Invocation.method(
          #dispose,
          [],
        ),
        returnValueForMissingStub: null,
      );
  @override
  void notifyListeners() => super.noSuchMethod(
        Invocation.method(
          #notifyListeners,
          [],
        ),
        returnValueForMissingStub: null,
      );
}
```
Might be an issue in analyzer 5.2.0?

**Comment by yanok** [comment]
@jpnurmi Thanks, that was not an analyzer issue, but your finding that downgrading to 5.2.0 allows to reproduce it helped me a lot. Turns out that was `package:build_resolvers` issue, fixed in 2.1.0. I bet that was the last item here: https://pub.dev/packages/build_resolvers/changelog#210

@polina-c you should be able to revert your fix and use `VoidCallback` directly again (provided `package:build_resolvers` is >=2.1.0)

**Comment by yanok** [comment]
Closing, as there is no feedback and I'm pretty confident it's a manifestation of the `build_resolvers` issue.
